main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala

```diff
@@ -0,0 +1,354 @@
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur
+
+object OccurrenceAnalyzer {
+
+
+  def run(root: LiftedAst.Root)(implicit flix: Flix): Validation[OccurrenceAst.Root, CompilationMessage] = flix.phase("OccurrenceAnalyzer") {
+    // TODO: Implement occurrence analysis.
+    val defs = root.defs.map {
+      case (sym, defn) =>
+        val fparams = defn.fparams.map { case LiftedAst.FormalParam(sym, mod, tpe, loc) => OccurrenceAst.FormalParam(sym, mod, tpe, loc) }
+        val (e, _) = visitExp(defn.exp)
+        sym -> OccurrenceAst.Def(defn.ann, defn.mod, defn.sym, fparams, e, defn.tpe, defn.loc)
+    }
+    val enums = root.enums.map {
+      case (sym, enum) =>
+        val cases = enum.cases.map {
+          case (tag, caze) =>
+            tag -> OccurrenceAst.Case(caze.sym, tag, caze.tpeDeprecated, caze.loc)
+        }
+        sym -> OccurrenceAst.Enum(enum.mod, enum.sym, cases, enum.tpeDeprecated, enum.loc)
+    }
+    // Reassemble the ast root.
+    val result = OccurrenceAst.Root(defs, enums, root.reachable, root.sources)
+    return result.toSuccess
+  }
+
+  def visitExp(exp0: LiftedAst.Expression): (OccurrenceAst.Expression, Map[Symbol.VarSym, OccurrenceAst.Occur]) = exp0 match {
+    case Expression.Unit(loc) => (OccurrenceAst.Expression.Unit(loc), Map.empty)
+
+    case Expression.Null(tpe, loc) => (OccurrenceAst.Expression.Null(tpe, loc), Map.empty)
+
+    case Expression.True(loc) => (OccurrenceAst.Expression.True(loc), Map.empty)
+
+    case Expression.False(loc) => (OccurrenceAst.Expression.False(loc), Map.empty)
+
+    case Expression.Char(lit, loc) => (OccurrenceAst.Expression.Char(lit, loc), Map.empty)
+
+    case Expression.Float32(lit, loc) => (OccurrenceAst.Expression.Float32(lit, loc), Map.empty)
+
+    case Expression.Float64(lit, loc) => (OccurrenceAst.Expression.Float64(lit, loc), Map.empty)
+
+    case Expression.Int8(lit, loc) => (OccurrenceAst.Expression.Int8(lit, loc), Map.empty)
+
+    case Expression.Int16(lit, loc) => (OccurrenceAst.Expression.Int16(lit, loc), Map.empty)
+
+    case Expression.Int32(lit, loc) => (OccurrenceAst.Expression.Int32(lit, loc), Map.empty)
+
+    case Expression.Int64(lit, loc) => (OccurrenceAst.Expression.Int64(lit, loc), Map.empty)
+
+    case Expression.BigInt(lit, loc) => (OccurrenceAst.Expression.BigInt(lit, loc), Map.empty)
+
+    case Expression.Str(lit, loc) => (OccurrenceAst.Expression.Str(lit, loc), Map.empty)
+
+    case Expression.Var(sym, tpe, loc) => (OccurrenceAst.Expression.Var(sym, tpe, loc), Map(sym -> Once))
+
+    case Expression.Closure(sym, freeVars, tpe, loc) =>
+      val freeVars1 = freeVars.map {
+        case LiftedAst.FreeVar(sym, tpe) => OccurrenceAst.FreeVar(sym, tpe)
+      }
+      (OccurrenceAst.Expression.Closure(sym, freeVars1, tpe, loc), Map.empty)
+
+    case Expression.ApplyClo(exp, args, tpe, loc) =>
+      val (e, o1) = visitExp(exp)
+      val (as, o2) = visitExps(args)
+      val m = merge(o1, o2)
```

Comment:
Here you use the name `m`, but in other places you use `o3` and in other places `bs`. Find one convention and stick to it :)